common/src/main/java/com/wynntils/core/functions/FunctionManager.java

```diff
@@ -99,132 +105,115 @@ private Optional<Object> getFunctionValueSafely(Function<?> function, String arg
+                return decimalFormat.format(number);
+            }
+        }
+
         return value.toString();
     }
 
-    /**
-     * Return a string, based on the template, with values filled in from the referenced
-     * functions.
-     */
-    public Component getStringFromTemplate(String template) {
-        // FIXME: implement template parser
-        return Component.literal(template);
+    // endregion
+
+    // region Raw value calculations
+    // These are needed for getting a function value without converting its type to a string
+
+    public ErrorOr<Object> getRawFunctionValue(Function<?> function, FunctionArguments arguments) {
+        Optional<Object> value = getFunctionValueSafely(function, arguments);
+        return value.map(ErrorOr::of)
+                .orElseGet(() -> ErrorOr.error("Failed to get value of function: " + function.getName()));
     }
 
-    /**
-     * Return a list of all functions referenced in a template string
-     */
-    public List<Function<?>> getFunctionsInTemplate(String template) {
-        // FIXME: implement template parser
-        return List.of();
+    // endregion
+
+    // region Template formatting
+
+    private String doFormat(String templateString) {
+        return TemplateParser.doFormat(templateString);
     }
 
-    public <T> void doFormat(
-            String format,
-            Consumer<T> consumer,
-            java.util.function.Function<String, T> mapper,
-            Map<String, T> infoVariableMap) {
-        Set<String> infoVariables = infoVariableMap.keySet();
-
-        int index = 0;
-        // TODO: Can we get away with less calculations since we now have asymmetric delimiters?
-        while (index < format.length()) {
-            int indexStartOfNextVariable = format.indexOf('{', index);
-            if (indexStartOfNextVariable == -1) {
-                break;
-            }
+    public String[] doFormatLines(String templateString) {
+        StringBuilder resultBuilder = new StringBuilder();
 
-            int indexEndOfNextVariable = format.indexOf('}', indexStartOfNextVariable + 1);
-            if (indexEndOfNextVariable == -1) {
-                break;
-            }
+        // Iterate though the string and escape characters
+        // that are prefixed with `\`, remove the `\` and add it to the result
 
-            if (index != indexStartOfNextVariable) { // update none done too
-                consumer.accept(mapper.apply(format.substring(index, indexStartOfNextVariable)));
-            }
+        for (int i = 0; i < templateString.length(); i++) {
+            char c = templateString.charAt(i);
+            if (c == '\\') {
+                if (i + 1 < templateString.length()) {
+                    char nextChar = templateString.charAt(i + 1);
 
-            String toMatch = format.substring(indexStartOfNextVariable + 1, indexEndOfNextVariable);
+                    resultBuilder.append(doEscapeFormat(nextChar));
```

Comment:
I tend to agree with Qodana here. Assigning to the loop variable is not nice. I know I have already approved this so I can look the other way and let you push this. But if you think it is possible to get away without doing this, it would be nice. This is a quite complex loop; it is possible it will be just worse if you try to do it. But otoh, just since it is complex, it helps to keep it simple by e.g. not violating commonly held assumptions about loop variables.

Now I said it at least. You decide if it is possible to make this cleaner, or if you just want to press Resolved and get on with your life. :)

Reply:
I don't think anything good will come out of this. I share your concern, but I think we might have to do it here.